src/main/java/io/temporal/samples/encryptedpayloads/CryptDataConverter.java

```diff
@@ -0,0 +1,189 @@
+  }
+
+  private byte[] encrypt(byte[] plainData) throws Exception {
+    SecretKey key = this.getKey();
+    byte[] nonce = getNonce(GCM_NONCE_LENGTH_BYTE);
+
+    Cipher cipher = Cipher.getInstance(CIPHER);
+    cipher.init(Cipher.ENCRYPT_MODE, key, new GCMParameterSpec(GCM_TAG_LENGTH_BIT, nonce));
+
+    byte[] encryptedData = cipher.doFinal(plainData);
+    byte[] result = ByteBuffer.allocate(nonce.length + encryptedData.length).put(nonce).put(encryptedData).array();
+
+    return result;
+  }
+
+  private byte[] decrypt(byte[] encryptedDataWithNonce) throws Exception {
+    SecretKey key = this.getKey();
+    ByteBuffer buffer = ByteBuffer.wrap(encryptedDataWithNonce);
+
+    byte[] nonce = new byte[GCM_NONCE_LENGTH_BYTE];
+    buffer.get(nonce);
+    byte[] encryptedData = new byte[buffer.remaining()];
+    buffer.get(encryptedData);
+
+    Cipher cipher = Cipher.getInstance(CIPHER);
+    cipher.init(Cipher.DECRYPT_MODE, key, new GCMParameterSpec(GCM_TAG_LENGTH_BIT, nonce));
+
+    return cipher.doFinal(encryptedData);
+  }
+
+  @Override
+  public <T> Optional<Payload> toPayload(T value) throws DataConverterException {
+    Optional<Payload> optionalPayload = converter.toPayload(value);
+
+    if (!optionalPayload.isPresent()) {
+      return optionalPayload;
+    }
+
+    Payload innerPayload = optionalPayload.get();
+
+    ByteString encoding = innerPayload.getMetadataOrDefault(EncodingKeys.METADATA_ENCODING_KEY, null);
+    if (encoding == null) {
+      return optionalPayload;
+    }
+
+    byte[] encryptedData;
+    try {
+      encryptedData = encrypt(innerPayload.getData().toByteArray());
+    } catch (Throwable e) {
+      throw new DataConverterException(e);
+    }
+
+    Payload encryptedPayload = Payload.newBuilder(innerPayload)
+      .setData(ByteString.copyFrom(encryptedData))
+      .putMetadata(EncodingKeys.METADATA_ENCODING_KEY, EncodingKeys.METADATA_ENCODING_RAW)
+      .putMetadata(EncodingKeys.METADATA_WRAPPED_ENCODING_KEY, encoding)
+      .build();
+
+    return Optional.of(encryptedPayload);
+  }
+
+  @Override
+  public <T> T fromPayload(Payload payload, Class<T> valueClass, Type valueType) {
+    ByteString encoding = payload.getMetadataOrDefault(EncodingKeys.METADATA_WRAPPED_ENCODING_KEY, null);
+    if (encoding == null) {
+      return converter.fromPayload(payload, valueClass, valueType);
+    }
+
+    byte[] plainData;
+    try {
+      plainData = decrypt(payload.getData().toByteArray());
+    } catch (Throwable e) {
+      throw new DataConverterException(e);
+    }
+
+    Payload decryptedPayload = Payload.newBuilder(payload)
+      .setData(ByteString.copyFrom(plainData))
+      .removeMetadata(EncodingKeys.METADATA_WRAPPED_ENCODING_KEY)
+      .putMetadata(EncodingKeys.METADATA_ENCODING_KEY, encoding)
+      .build();
+
+    return converter.fromPayload(decryptedPayload, valueClass, valueType);
+  }
+
+  @Override
+  public Optional<Payloads> toPayloads(Object... values) throws DataConverterException {
```

Comment:
I would call `converter.toPayloads(values)` and then serialize the resulting Payload (using protobuf generated method) and then encrypt the whole thing at once.

Reply:
This didn't quite feel safe to me as Payloads has no metadata. I have however taken the suggestion of encrypting/decrypting the serialised payload rather than encrypting only the data. This is much cleaner now :)